channeldb/payment_control_test.go

```diff
@@ -768,6 +768,302 @@ func TestPaymentControlDeletePayments(t *testing.T) {
+	attemptID := uint64(0)
+	for i := 0; i < 4; i++ {
+		info, attempt, preimg, err := genInfo()
+		if err != nil {
+			t.Fatalf("unable to generate htlc message: %v", err)
+		}
+
+		paymentIDs[i] = info.PaymentIdentifier
+
+		attempt.AttemptID = attemptID
+		attemptID++
+
+		// Init the payment.
+		err = pControl.InitPayment(info.PaymentIdentifier, info)
+		if err != nil {
+			t.Fatalf("unable to send htlc message: %v", err)
+		}
+
+		// Register and fail the first attempt for all three payments.
+		_, err = pControl.RegisterAttempt(info.PaymentIdentifier, attempt)
+		if err != nil {
+			t.Fatalf("unable to send htlc message: %v", err)
+		}
+
+		htlcFailure := HTLCFailUnreadable
+		_, err = pControl.FailAttempt(
+			info.PaymentIdentifier, attempt.AttemptID,
+			&HTLCFailInfo{
+				Reason: htlcFailure,
+			},
+		)
+		if err != nil {
+			t.Fatalf("unable to fail htlc: %v", err)
+		}
+
+		// Depending on the test case, fail or succeed the next
+		// attempt.
+		attempt.AttemptID = attemptID
+		attemptID++
+
+		_, err = pControl.RegisterAttempt(info.PaymentIdentifier, attempt)
+		if err != nil {
+			t.Fatalf("unable to send htlc message: %v", err)
+		}
+
+		switch i {
+
+		// Fail the attempt and the payment overall.
+		case 0, 1:
+			htlcFailure := HTLCFailUnreadable
+			_, err = pControl.FailAttempt(
+				info.PaymentIdentifier, attempt.AttemptID,
+				&HTLCFailInfo{
+					Reason: htlcFailure,
+				},
+			)
+			if err != nil {
+				t.Fatalf("unable to fail htlc: %v", err)
+			}
+
+			failReason := FailureReasonNoRoute
+			_, err = pControl.Fail(info.PaymentIdentifier, failReason)
+			if err != nil {
+				t.Fatalf("unable to fail payment hash: %v", err)
+			}
+
+		// Settle the attempt
+		case 2:
+			_, err := pControl.SettleAttempt(
+				info.PaymentIdentifier, attempt.AttemptID,
+				&HTLCSettleInfo{
+					Preimage: preimg,
+				},
+			)
+			if err != nil {
+				t.Fatalf("error shouldn't have been received, got: %v", err)
+			}
+
+		// We leave the attmpet in-flight by doing nothing.
+		case 3:
+		}
+	}
+
+	type fetchedPayment struct {
+		id     lntypes.Hash
+		status PaymentStatus
+		htlcs  int
+	}
+
+	assertPayments := func(expPayments []fetchedPayment) {
+		t.Helper()
+
+		dbPayments, err := db.FetchPayments()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if len(dbPayments) != len(expPayments) {
+			t.Fatalf("expected %d payments, got %d",
+				len(expPayments), len(dbPayments))
+		}
+
+		for i := range dbPayments {
+			if dbPayments[i].Info.PaymentIdentifier != expPayments[i].id {
+				t.Fatalf("unexpected payment identifier")
+			}
+
+			if dbPayments[i].Status != expPayments[i].status {
+				t.Fatalf("unexpected payment status: ID: %v, status: %v=%v", i, dbPayments[i].Status, expPayments[i].status)
+			}
+
+			if len(dbPayments[i].HTLCs) != expPayments[i].htlcs {
+				t.Fatalf("unexpected number of htlcs")
+			}
+		}
+	}
+
+	// Check that all payments are there as we added them.
+	assertPayments([]fetchedPayment{
```

Comment:
Could create this slice once initially and also use it as a template to create the payments (using the status instead of the index to determine what test data to store. Then this could be refactored into `assertPayments(0, 1, 2, 3)` to make sure the expected payments with the given IDs are still being found.